File: myapp/src/main/java/com/app/myapp/MainActivity.java
package com.app.myapp;

import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.view.Menu;
import android.view.MenuItem;

import com.app.myapp.dex.OtherApkActivity;
import com.app.myapp.pack.ui.PackManagerActivity;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        FloatingActionButton fab = findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                        .setAction("Action", null).show();
            }
        });

        findViewById(R.id.bt_package).setOnClickListener(this);
        findViewById(R.id.bt_other_apk).setOnClickListener(this);

    }

    @Override
    public void onClick(View view) {
        if(view.getId() == R.id.bt_package){
            startActivity(new Intent(MainActivity.this, PackManagerActivity.class));
        }else if(view.getId() == R.id.bt_other_apk){
            startActivity(new Intent(MainActivity.this, OtherApkActivity.class));
        }
    }
}

File: kot/src/main/java/com/app/kot/MainActivity.kt
package com.app.kot

import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.View.TEXT_ALIGNMENT_GRAVITY
import android.view.View.TEXT_DIRECTION_FIRST_STRONG
import kotlinx.android.synthetic.main.activity_main.*
import kotlin.properties.Delegates

class MainActivity : AppCompatActivity() {


    var test by Delegates.observable(0){property, oldValue, newValue ->

        println("====="+newValue)

    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)

        fab.setOnClickListener { view ->
            Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                    .setAction("Action", null).show()
        }

        var a = A("12")
        a.name

//        a.setClickListener(object : B {
//            override fun getName(name: String) {
//                println("======="+name)
//            }
//        })

//        a.setClickListener(object : OnNameListener{
//            override fun onNameDone(name: String) {
//
//            }
//        })
//        a.setClickListener{ i: Int, s: String ->
//            Log.i("========",i.toString())
//            Log.i("========",s)
//        }

        a.setSimpleCallBack {
            onSuccessFun {
                response ->
                Log.i("========",response)
            }
            
            onFailedFun { errorCode, msg ->
                Log.i("========",errorCode.toString()+"=="+msg)
            }

        }
        a.setClickListener { a, b ->

        }

//        A("123").apply {
//            setClickListener(object :B{
//                override fun getName(name: String) {
//
//                }
//            })
//            setTest()
//        }



//        with(fab){
//            textAlignment = TEXT_ALIGNMENT_GRAVITY
//            textDirection = TEXT_DIRECTION_FIRST_STRONG
//        }
//        fab.let {
//
//        }

        var p = plus(1,2){
            a,b -> a+b
        }



        //Log.i("111111111111",p.toString())
        testFun {
            2
        }
        

    }

    fun plus(num1 :Int,num2:Int,result:(a:Int,b:Int)->Int) : Int{
        return result(num1,num2)
    }


    fun testFun(listener:(String) ->Int){
        var a = listener("123")
        Log.i("=======1",a.toString())
    }

    enum class Test{
        RED,CL
    }
}
